### PR TITLE
Add support for generics

### PR DESCRIPTION
**Pull request by jjedelsky**
This pull adds support for generics.
See tests for examples.

### CODE REVIEW

**Comment by Arnavion** [comment]
@Yamakaky error-chain doesn't support generic errorkind, right?

**Comment by jjedelsky** [comment]
That is right, it doesn't.

File: derive-error-chain-tests/src/main.rs [hunk 1]
```diff
@@ -269,6 +269,66 @@ mod attributes_test {
 	}
 }
 
+mod generics_test {
```
Comment (Arnavion) [inline derive-error-chain-tests/src/main.rs]: Move this to the end of the file since it belongs in the "Own tests" section.

File: derive-error-chain/src/lib.rs [hunk 2]
```diff
@@ -640,14 +671,15 @@ This struct is made of three things:
 			", error_kind_name);
 
 			let result_wrapper = result_name.map(|result_name| quote! {
+				#[allow(non_camel_case_types)]
 				/// Convenient wrapper around `::std::result::Result`
-				pub type #result_name<T> = ::std::result::Result<T, #error_name>;
+				pub type #result_name#ty_generics_ext1 = ::std::result::Result<DERIVE_ERROR_CHAIN_T, #error_name#ty_generics>;
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: Space between the tokens. `#result_name #ty_generics_ext1`, `#error_name #ty_generics`, etc. There are some more lines that also need it.

File: derive-error-chain/src/lib.rs [hunk 3]
```diff
@@ -793,19 +825,20 @@ This struct is made of three things:
 					#extract_backtrace_fn
 				}
 
+				#[allow(non_camel_case_types)]
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: Why is this needed now?

File: derive-error-chain/src/lib.rs [hunk 4]
```diff
@@ -258,6 +258,37 @@ pub fn derive_error_chain(input: proc_macro::TokenStream) -> proc_macro::TokenSt
 	let ast = syn::parse_derive_input(&source).unwrap();
 	let error_kind_name = ast.ident;
 
+	let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+	let mut generics_lifetime = ast.generics.clone();
+	generics_lifetime.lifetimes.push(syn::LifetimeDef::new("'error_chain_lifetime"));
+	let (impl_generics_lifetime, _, _) = generics_lifetime.split_for_impl();
+	let mut generics_ext1 = ast.generics.clone();
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: `generics_ext1` -> `result_generics`

File: derive-error-chain/src/lib.rs [hunk 5]
```diff
@@ -258,6 +258,37 @@ pub fn derive_error_chain(input: proc_macro::TokenStream) -> proc_macro::TokenSt
...
+	generics_ext1.ty_params.push(syn::TyParam {
+		attrs: Vec::new(),
+		ident: syn::Ident::from("DERIVE_ERROR_CHAIN_T"),
+		bounds: Vec::new(),
+		default: None,
+	});
+	let (_, ty_generics_ext1, _) = generics_ext1.split_for_impl();
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: `ty_generics_ext1` -> `result_ty_generics`

File: derive-error-chain/src/lib.rs [hunk 6]
```diff
@@ -258,6 +258,37 @@ pub fn derive_error_chain(input: proc_macro::TokenStream) -> proc_macro::TokenSt
...
+	let mut generics_ext2 = ast.generics.clone();
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: `generics_ext2` -> `result_ext_generics`

File: derive-error-chain/src/lib.rs [hunk 7]
```diff
@@ -258,6 +258,37 @@ pub fn derive_error_chain(input: proc_macro::TokenStream) -> proc_macro::TokenSt
...
+		attrs: Vec::new(),
+		ident: syn::Ident::from("DERIVE_ERROR_CHAIN_E"),
+		bounds: vec![
+			syn::parse_ty_param_bound("::std::error::Error").unwrap(),
+			syn::parse_ty_param_bound("::std::marker::Send").unwrap(),
+			syn::parse_ty_param_bound("'static").unwrap(),
+		],
+		default: None,
+	});
+	let (impl_generics_ext2, ty_generics_ext2, _) = generics_ext2.split_for_impl();
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: `impl_generics_ext2` -> `result_ext_impl_generics`

`ty_generics_ext2` -> `result_ext_ty_generics`

File: derive-error-chain/src/lib.rs [hunk 8]
```diff
@@ -793,19 +825,20 @@ This struct is made of three things:
...
-						where F: FnOnce() -> EK, EK: Into<#error_kind_name>;
+					fn chain_err<DERIVE_ERROR_CHAIN_F, DERIVE_ERROR_CHAIN_EK>(self, callback: DERIVE_ERROR_CHAIN_F) -> ::std::result::Result<DERIVE_ERROR_CHAIN_T, #error_name#ty_generics>
+						where DERIVE_ERROR_CHAIN_F: FnOnce() -> DERIVE_ERROR_CHAIN_EK, DERIVE_ERROR_CHAIN_EK: Into<#error_kind_name#ty_generics>;
 				}
 
-				impl<T, E> #result_ext_name<T, E> for ::std::result::Result<T, E>
-					where E: ::std::error::Error + Send + 'static {
-					fn chain_err<F, EK>(self, callback: F) -> ::std::result::Result<T, #error_name>
-						where F: FnOnce() -> EK, EK: Into<#error_kind_name> {
+				#[allow(non_camel_case_types)]
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: Same here. Why is this needed now?

File: derive-error-chain-tests/src/main.rs [hunk 9]
```diff
@@ -269,6 +269,66 @@ mod attributes_test {
...
+		#[error_chain(display = r#"|t| write!(f, "custom generic boxed error: {}", t)"#)]
+		CustomGenericBoxed(Box<U>),
+
+		#[error_chain(link = "inner1::Error<U>")]
+		LinkGeneric(inner1::ErrorKind<U>),
+
+		#[error_chain(link = "inner2::Error")]
+		Link(inner2::ErrorKind),
+
+		// FIXME: conflicting implementations of trait `std::convert::From<&str>` for type `generics_test::Error<&str, _>
```
Comment (Arnavion) [inline derive-error-chain-tests/src/main.rs]: Is this accurate? AFAICT this would only happen if `T` was a `String` or `&str`, but that can't be because it both violates the semantics of foreign errors and doesn't satisfy the `T: Error` generic bound anyway.

File: derive-error-chain/src/lib.rs [hunk 10]
```diff
@@ -258,6 +258,37 @@ pub fn derive_error_chain(input: proc_macro::TokenStream) -> proc_macro::TokenSt
...
+		attrs: Vec::new(),
+		ident: syn::Ident::from("DERIVE_ERROR_CHAIN_T"),
+		bounds: Vec::new(),
+		default: None,
+	});
+	let (_, ty_generics_ext1, _) = generics_ext1.split_for_impl();
+	let mut generics_ext2 = ast.generics.clone();
+	generics_ext2.ty_params.push(syn::TyParam {
+		attrs: Vec::new(),
+		ident: syn::Ident::from("DERIVE_ERROR_CHAIN_T"),
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: I don't know of any precedent in custom derives about naming generics that don't clash with generics in the input, but perhaps `__T`, `__E`, `__F`, `__K` would be better?

**Review by Arnavion — CHANGES_REQUESTED** [review]
I've just eyeballed it and left some comments. I'll properly test it on the weekend.

**Comment by Arnavion** [comment]
Right, so error_chain errorkinds would no longer be guaranteed to be able to hold derive_error_chain errorkinds as chainable links. (Not a blocker, but worth noting.)

File: derive-error-chain-tests/src/main.rs [hunk 11]
```diff
@@ -511,8 +511,11 @@ mod generics_test {
 		#[error_chain(link = "inner2::Error")]
 		Link(inner2::ErrorKind),
 
-		// FIXME: conflicting implementations of trait `std::convert::From<&str>` for type `generics_test::Error<&str, _>
-		// FIXME: conflicting implementations of trait `std::convert::From<std::string::String>` for type `generics_test::Error<std::string::String, _>
+		// FIXME: conflicting implementations of trait `std::convert::From<generics_test::inner1::Error<_>>` for type `generics_test::Error<generics_test::inner1::Error<_>, _>`
```
Comment (Arnavion) [inline derive-error-chain-tests/src/main.rs]: Indeed, the conflict is between the From impl for LinkGeneric and ForeignGeneric.

```rust
impl <T: error::Error + Send + 'static, U> From<inner1::Error<U>> for Error<T, U>
    where U: Send + fmt::Debug + fmt::Display + 'static {

    fn from(err: inner1::Error<U>) -> Self {
        Error(ErrorKind::LinkGeneric(err.0), err.1)
    }
}
```

and

```rust
impl <T: error::Error + Send + 'static, U> From<T> for Error<T, U>
    where U: Send + fmt::Debug + fmt::Display + 'static {

    fn from(err: T) -> Self {
        Self::from_kind(ErrorKind::ForeignGeneric(err))
    }
}
```

File: derive-error-chain-tests/src/main.rs [hunk 12]
```diff
@@ -511,8 +511,11 @@ mod generics_test {
...
+		// FIXME: conflicting implementations of trait `std::convert::From<generics_test::inner2::Error>` for type `generics_test::Error<generics_test::inner2::Error, _>`
```
Comment (Arnavion) [inline derive-error-chain-tests/src/main.rs]: Indeed, the conflict is between the From impl for Link and ForeignGeneric

```rust
impl <T: error::Error + Send + 'static, U> From<inner2::Error> for Error<T, U>
    where U: Send + fmt::Debug + fmt::Display + 'static {

    fn from(err: inner2::Error) -> Self {
        Error(ErrorKind::Link(err.0), err.1)
    }
}
```

and

```rust
impl <T: error::Error + Send + 'static, U> From<T> for Error<T, U>
    where U: Send + fmt::Debug + fmt::Display + 'static {

    fn from(err: T) -> Self {
        Self::from_kind(ErrorKind::ForeignGeneric(err))
    }
}
```

File: derive-error-chain-tests/src/main.rs [hunk 13]
```diff
@@ -511,8 +511,11 @@ mod generics_test {
...
+		// FIXME: conflicting implementations of trait `std::convert::From<std::io::Error>` for type `generics_test::Error<std::io::Error, _>`
+		// FIXME: conflicting implementations of trait `std::convert::From<&str>` for type `generics_test::Error<&str, _>`
```
Comment (Arnavion) [inline derive-error-chain-tests/src/main.rs]: Ah, so the issue for this (and for String) below is that the compiler seems to ignore the `std::error::Error` bound on `T` when determining if conflicts exist.

File: derive-error-chain/src/lib.rs [hunk 14]
```diff
@@ -258,6 +258,37 @@ pub fn derive_error_chain(input: proc_macro::TokenStream) -> proc_macro::TokenSt
 	let ast = syn::parse_derive_input(&source).unwrap();
 	let error_kind_name = ast.ident;
 
+	let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+	let mut generics_lifetime = ast.generics.clone();
+	generics_lifetime.lifetimes.push(syn::LifetimeDef::new("'__a"));
+	let (impl_generics_lifetime, _, _) = generics_lifetime.split_for_impl();
+	let mut result_generics = ast.generics.clone();
+	result_generics.ty_params.push(syn::TyParam {
+		attrs: Vec::new(),
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: All `Vec::new()` -> `vec![]`

File: derive-error-chain/src/lib.rs [hunk 15]
```diff
@@ -640,14 +671,15 @@ This struct is made of three things:
 			", error_kind_name);
 
 			let result_wrapper = result_name.map(|result_name| quote! {
+				#[allow(non_camel_case_types)]
```
Comment (Arnavion) [inline derive-error-chain/src/lib.rs]: Remove this.

**Review by Arnavion — CHANGES_REQUESTED** [review]
Looks good to me modulo two small changes and your FIXMEs. What do you plan to do about the FIXMEs?

**Comment by jjedelsky** [comment]
I thing it should be enough not to implement `From<generic> for Error`.

**Comment by Arnavion** [comment]
Thanks!

I made some tiny changes:

- Removed the import for `std::collections::HashSet` since I prefer to use fully-qualified paths.
- Changed the code that builds the set to use `collect()` instead of a `for`-loop.
- Moved the generic-matcher to its own match arm.
- Added a check to the generic-matcher that the type path has only one segment so that it doesn't match `some_mod::T`.
- Some comments.
